sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala

```diff
@@ -130,6 +143,202 @@ class BulkWriter(container: CosmosAsyncContainer,
+                      // But even the above approach can only work within the same batch but not for the whole spark partition processing.
+                      for (readManyOperation <- readManyOperations) {
+                          val cosmosPatchBulkUpdateOperations =
+                              cosmosPatchHelperOpt
+                                  .get
+                                  .createCosmosPatchBulkUpdateOperations(readManyOperation.objectNode)
+
+                          val rootNode =
+                              cosmosPatchHelperOpt
+                                  .get
+                                  .patchBulkUpdateItem(resultMap.get(readManyOperation.cosmosItemIdentity), cosmosPatchBulkUpdateOperations)
+
+                          // create bulk item operation
+                          val etagOpt = Option.apply(rootNode.get(CosmosConstants.Properties.ETag))
+                          val bulkItemOperation = etagOpt match {
+                              case Some(etag) =>
+                                  CosmosBulkOperations.getReplaceItemOperation(
+                                      readManyOperation.cosmosItemIdentity.getId,
+                                      rootNode,
+                                      readManyOperation.cosmosItemIdentity.getPartitionKey,
+                                      new CosmosBulkItemRequestOptions().setIfMatchETag(etag.asText()),
+                                      OperationContext(
+                                          readManyOperation.operationContext.itemId,
+                                          readManyOperation.operationContext.partitionKeyValue,
+                                          Some(etag.asText()),
+                                          readManyOperation.operationContext.attemptNumber,
+                                          Some(readManyOperation.objectNode)
+                                      ))
+                              case None => CosmosBulkOperations.getCreateItemOperation(
+                                  rootNode,
+                                  readManyOperation.cosmosItemIdentity.getPartitionKey,
+                                  OperationContext(
+                                      readManyOperation.operationContext.itemId,
+                                      readManyOperation.operationContext.partitionKeyValue,
+                                      None,
+                                      readManyOperation.operationContext.attemptNumber,
+                                      Some(readManyOperation.objectNode)
+                                  ))
+                          }
+
+                          this.emitBulkInput(bulkItemOperation)
+                      }
+                  })
+                  .onErrorResume(throwable => {
+                      for (readManyOperation <- readManyOperations) {
```

Comment:
QUESTION: This error handling seems fine - but wouldn't we also need ot make sure that for certain errors 409 on create and 412 on Replace the retry would redo the read-many as well?

Reply:
What i mean is I saw that you added the error handling on patch in shoudlRetry for 409/412 - but the errors wouldn't happen / be caught here - because you just schedule the bulk operation?

Reply:
yea we will not get 409 or 412 in readMany, it will only happen during bulk write.

in the shouldRetry logic, it added the 409 and 412 checking for patchBulkUpdate.  so when an operation failed during the bulkWrite step and when we retry, it will go through the whole flow again -> first readmany then bulkWrite

Reply:
Got it - yes, should work as is